lib/libc/sys/epoll.2
.\"	$NetBSD: epoll.2,v 1.2 2023/07/28 23:41:16 wiz Exp $
.\"
.\" Copyright (c) 2023 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to The NetBSD Foundation
.\" by Theodore Preduta.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd July 19, 2023
.Dt EPOLL 2
.Os
.Sh NAME
.Nm epoll ,
.Nm epoll_event ,
.Nm epoll_data ,
.Nm epoll_data_t ,
.Nm epoll_create ,
.Nm epoll_create1 ,
.Nm epoll_ctl ,
.Nm epoll_wait ,
.Nm epoll_pwait ,
.Nm epoll_pwait2
.Nd event notification mechanism
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In sys/epoll.h
.Bd -literal
union epoll_data {
	void		*ptr;
	int		fd;
	uint32_t	u32;
	uint64_t	u64;
};

typedef union epoll_data	epoll_data_t;

struct epoll_event {
	uint32_t	events;
	epoll_data_t	data;
};
.Ed
.Pp
.Ft int
.Fn epoll_create "int size"
.Ft int
.Fn epoll_create1 "int flags"
.Ft int
.Fn epoll_ctl "int epfd" "int op" "int fd" "struct epoll_event *event"
.Ft int
.Fn epoll_wait "int epfd" "struct epoll_event *events" "int maxevents" "int timeout"
.Ft int
.Fn epoll_pwait "int epfd" "struct epoll_event *events" "int maxevents" "int timeout" "const sigset_t *sigmask"
.Ft int
.Fn epoll_pwait2 "int epfd" "struct epoll_event *events" "int maxevents" "const struct timespec *timeout" "const sigset_t *sigmask"
.Sh DESCRIPTION
.Nm
provides a similar facility to both
.Xr select 2
and
.Xr kqueue 2 :
it allows for the examination of file descriptors to see if they are available
for reading/writing.
.Pp
The
.Va epoll_event
structure consists of two fields,
.Va events
and
.Va data .
The
.Va data
field is passed through the kernel and is intended to be used to identify the
event.
When used with
.Fn epoll_ctl ,
the
.Va events
field consists of a mask of the events that the
.Nm
instance should watch for, and when being used with
.Fn epoll_wait ,
.Fn epoll_pwait ,
and
.Fn epoll_pwait2
consists of a mask of the events that occurred.
The following are possible values for
.Va events :
.Bl -tag -width EPOLLONESHOT
.It Dv EPOLLIN
Watch for
.Xr read 2
operations.
.It Dv EPOLLOUT
Watch for
.Xr write 2
operations.
.It Dv EPOLLRDHUP
Watch for a peer closed connection.
.It Dv EPOLLERR
Watch for error conditions.
.It Dv EPOLLET
This option modifies the other set bits of
.Va events .
When set, the events described by other bits in
.Va events
are only triggered when the state change.
Otherwise the events are considered triggered whenever the condition is true.
.It Dv EPOLLONESHOT
Remove this event once it is retrieved once.
.El
.Pp
.Fn epoll_create
and
.Fn epoll_create1
both create an
.Nm
instance.
The
.Fa size
argument specified for
.Fn epoll_create
exists so that the
.Nx
function has the same signature as the Linux system call of the same name.
.Fa size
must be positive, but is otherwise unused.
Additionally, optionally,
.Dv EPOLL_CLOEXEC
may be specified in the
.Fa flags
of
.Fn epoll_create1
to set the
.Xr close 2
on
.Xr exec 3
flag.
.Pp
.Fn epoll_ctl
is used to make changes to the given
.Nm
instance based on the provided
.Fa op .
Possible values for
.Fa op
are:
.Bl -tag -width EPOLL_CTL_ADD
.It Dv EPOLL_CTL_ADD
Register interest of
.Fa fd
on the
.Fa epfd
for the events specified in
.Fa event .
.It Dv EPOLL_CTL_MOD
Modify the events registered for
.Fa fd
to those specified in
.Fa event .
.It Dv EPOLL_CTL_DEL
Deregister
.Fa fd
from the
.Nm
instance specified in
.Fa epfd .
Note that
.Fa event
is completely ignored in this case.
.El
.Pp
.Fn epoll_wait ,
.Fn epoll_pwait ,
and
.Fn epoll_pwait2
provide the ability to wait for up to
.Fa maxevents
which are stored in the buffer pointed to by
.Fa events .
For
.Fn epoll_wait
and
.Fn epoll_pwait ,
a timeout may be specified in
.Fa timeout
in milliseconds.
If no timeout is desired, -1 should be specified in
.Fa timeout .
For
.Fn epoll_pwait2
if no timeout is desired
.Fa timeout
should be specified as
.Dv NULL .
Additionally,
a sigmask may be specified to
.Fn epoll_pwait
and
.Fn epoll_pwait2
in
.Fa sigmask
to set the sigmask while
.Nm
waits for events.
.Pp
Note that
.Nm
is not intended to be used by native
.Nx
applications.
Instead it is only intended to used as a means to help port software originally
written for Linux to
.Nx .
.Sh RETURN VALUES
.Fn epoll_create
and
.Fn epoll_create1
both return a file descriptor when successful.
.Pp
.Fn epoll_ctl
returns zero when successful.
.Pp
.Fn epoll_wait ,
.Fn epoll_pwait ,
and
.Fn epoll_pwait2
return the number of events written to
.Fa events
when successful.
Note that zero is written to when
.Fa timeout
expires and no events were available.
.Pp
When any of the above fail, -1 is returned and
.Fa errno
is set.
.Sh ERRORS
The
.Fn epoll_create
and
.Fn epoll_create1
functions fail if:
.Bl -tag -width Er
.It Bq Er EINVAL
.Fa size
is not positive.
.Pp
Bits other than
.Dv EPOLL_CLOEXEC
are provided in
.Fa flags .
.It Bq Er EMFILE
The per-process descriptor table is full.
.It Bq Er ENFILE
The system file table is full.
.It Bq Er ENOMEM
The kernel failed to allocate enough memory for a
.Nm
instance.
.El
.Pp
The
.Fn epoll_ctl
function fails if:
.Bl -tag -width Er
.It Bq Er EBADF
.Fa epfd
or
.Fa fd
is not a valid file descriptor.
.It Bq Er EEXIST
.Fa op
is
.Dv EPOLL_CTL_ADD
and
.Fa fd
was already previously added via
.Dv EPOLL_CTL_ADD .
.It Bq Er EINVAL
.Fa epfd
is not a file descriptor for an
.Nm
instance.
.Pp
.Fa epfd
and
.Fa fd
represent the same
.Nm
instance.
.It Bq Er ELOOP
.Fa op
is
.Dv EPOLL_CTL_ADD
and adding
.Fa fd
would result in a circular loop of
.Nm
instances.
.It Bq Er ENOENT
.Fa op
is
.Dv EPOLL_CTL_MOD
or
.Dv EPOLL_CTL_DEL
and
.Fa fd
was not previously added with
.Dv EPOLL_CTL_ADD .
.It Bq Er ENOMEM
The kernel failed to allocate enough memory for
.Fa op .
.It Bq Er EPERM
.Fa fd
does not support
.Nm epoll .
.El
.Pp
The
.Fn epoll_wait ,
.Fn epoll_pwait ,
and
.Fn epoll_pwait2
functions fail if:
.Bl -tag -width Er
.It Bq Er EBADF
.Fa epfd
is not a valid file descriptor.
.It Bq Er EFAULT
The area provided in
.Fa events
failed to be written to.
.It Bq Er EINTR
A signal was delivered before any events became available and
.Fa timeout
expired.
.It Bq Er EINVAL
.Fa epfd
is not a valid
.Nm
file descriptor.
.Pp
.Fa maxevents
is less than or equal to zero.
.El
.Sh SEE ALSO
.Xr kqueue 2 ,
.Xr poll 2 ,
.Xr select 2
.Sh HISTORY
The
.Nm
functions and types are designed to be compatible with the Linux system calls of
the same name.
.Sh CAVEATS
The
.Nm
facility is not intended to be used in conjunction with
.Xr kqueue 2 .
.Pp
Unlike Linux's
.Nm ,
the
.Nx
version does not survive a
.Xr fork 2 .


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.2989 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------

Changes from NetBSD 10.0 to NetBSD 11.0:
	amdsmn(4): added trivial power management support [reinoud 20221218]
	amdccp(4): added trivial power management support [reinoud 20221218]
	libc: Update Unicode tables to Unicode 13.0.0. [wiz 20221220]
	libc: Update Unicode tables to Unicode 14.0.0. [wiz 20221220]
	ppp(4): Remove several non-functioning ioctls. [chs 20221221]
	libc: Update Unicode tables to Unicode 15.0.0. [wiz 20221221]
	dse(4): Added Dayna SCSI/Link ethernet driver. [nat 20221222]
	binutils: Updated to FSF binutils 2.39. [christos 20221224]
	sdhc(4): Add quirk for some Intel eMMC devices (like
		GIGABYTE MA10-ST0) to allow the driver to attach
		successfully. [msaitoh 20230105]
	cribbage(6): Add -y option to keep playing without asking
		(from Greywolf in PR 57164) [christos 20230106]
	x86: Enable the -z separate-code security feature by default in
		ld(1). Reduces the default max page size from 2M to
		4K on amd64. [christos 20230106]
	bind: Import version 9.16.37. [christos 20230125]
	xfwp(1): remove. [wiz 20230204]
	wscons(4): Ignore nonsense tab stops in vt100 emulation.
		[riastradh 20230223]
	nvmm(4): Filter CR4 bits on x86 SVM (AMD). In particular,
		prohibit PKE, Protection Key Enable, which requires
		some additional management of CPU state by nvmm. 
		[riastradh 20230223]
	ftp(1): Add option sslnoverify to control validation of SSL
		certificates. Certificate validation is now enabled
		by default. [mlelstv 20230225]
	ftp(1): Add netrc processing to fetch-mode (URL on command line) to
		enable options and autologin via netrc. [mlelstv 20230225]
	aarch64: Audit use of curcpu(9). [riastradh 20230225]
	xen: Memory barrier audit. Reduce unnecessary membars.
		[riastradh 20230225]
	x86: Mitigate MXCSR Configuration Dependent Timing CPU security
		flaw in kernel FPU use. [riastradh 20230225]
	kernel: Optimization: Omit needless membar when triggering softint
		in various ports. [riastradh 20230301]
	sh(1): Adjust tilde expansion to comply with upcoming POSIX
		Issue 8 requirements. [kre 20230306]
	libm: Add more long double transcendental functions (as aliases of
		double functions) from C99. [riastradh 20230313]
	tzdata: updated to 2023a (using the 2023agtz fork) [kre 20230323]
	tzdata: updated to 2023b (using the 2023bgtz fork) [kre 20230324]
	xen: Allow a PVH dom0 to use VGA as console. [bouyer 20230324]
	vioscsi(4), viornd(4), viomb(4): Mark VirtIO drivers as MPSAFE.
		[mlelstv 20230325]
	amiga: Fix bootblocks for Kickstart 3.2. [abs 20230325]
	libc: Add __atomic_is_lock_free to help stdatomic.h usage with
		GCC. [riastradh 20230330]
	tzdata: updated to 2023c (using the 2023cgtz fork) [kre 20230331]
	tprof(8): Add Intel Comet Lake support. [msaitoh 20230410]
	vacation(1): Check 'Auto-Submitted:' header in addition to
		'Precedence:', and set 'Precedence:' in addition to
		'Auto-Submitted:'.  [hauke 20230411]
	ichsmb(4), tco(4): Add support for TCO on newer Intel chipsets.
		[riastradh 20230412]
	audiorecord(1): Add support for recording 24-bit wav files.
		[mlelstv 20230415]
	lua: Import version 5.4.4.
		Bumped shlib major because of incompatible API/ABI changes.
		[nikita 20230416]
	evbarm: Install a boot.cfg file for the EFI bootloader by default.
		[brook 20230419]
	dhcpcd: Import version 10.0.1. [roy 20230421]
	evbarm: Enable rkv1crypto driver on Rock64. [mrg 20230424]
	worms(6): Add a little bit of optional colour, code and
		ideas from RVP. [kre 20230426]
	amiga: Add driver for the MNT ZZ9000 Zorro board - submitted by
		Alain Runa. [phx 20230503]
	evbarm: Experimental IMX6SX support - tested on the UDOO Neo Full board.
		[bouyer 20230504]
	chown(8): Add a -d flag to avoid changing a file's owner/group to the
		current value.  This avoids some unnecessary operations on the
		file. [pgoyette 20230504]
	ftp(1): Add timeout for SSL connection setup, defaulting to 60 seconds.
		[lukem 20230505]
	riscv: Add RISC-V support that works on QEMU with a single hart.
		[skrll 20230507]
	riscv: Add Google Goldfish RTC driver. [skrll 20230508]
	openssl(1): Imported 3.0.8. [christos 20230510]
	make(1): Make :mtime operate on each word in variable
		value. [sjg 20230510]
	worms(6): When using -H, make the appearance of the worm head
		depend upon the body (more than it did before). [kre 20230512]
	mips: Bump the maximum text size for binaries from 64MB
		to 128MB. [he 20230514]
	amiga: Update loadbsd source and distribution binary to version 3.3.
		[phx 20230514]
	i386: Add /dev/efi device. [riastradh 20230522]
	efi(4): Add EFIIOC_GET_TABLE ioctl support on x86. [riastradh 20230522]
	ctfmerge(1): Add error handling to pthread APIs.
		Fix semaphores for macOS host tool.  [lukem 20230523]
	share/mk: Deprecate non-functional SHAREDSTRINGS. [lukem 20230524]
	mkstr(1), xstr(1): Obsolete unused PDP-11-era tools. [lukem 20230524]
	pthread(3): Reduce power consumption when spinning for a lock on
		arm. [riastradh 20230525]
	pax(1): Exit 0 if stdin filelist is used and no files are supplied.

lib/libc/sys/epoll.c
/*	$NetBSD: epoll.c,v 1.1 2023/07/28 18:19:00 christos Exp $	*/

/*-
 * Copyright (c) 2023 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
#include <sys/cdefs.h>
__RCSID("$NetBSD: epoll.c,v 1.1 2023/07/28 18:19:00 christos Exp $");

#include <sys/epoll.h>
#include <sys/sigtypes.h>
#include <sys/time.h>

#include <errno.h>
#include <stddef.h>

int
epoll_create(int size)
{
	if (size <= 0) {
		errno = EINVAL;
		return -1;
	}

	return epoll_create1(0);
}

int
epoll_wait(int epfd, struct epoll_event *events, int maxevents, int timeout)
{
	return epoll_pwait(epfd, events, maxevents, timeout, NULL);
}

int
epoll_pwait(int epfd, struct epoll_event *events, int maxevents, int timeout,
    const sigset_t *sigmask)
{
	struct timespec ts, *tsp;

	if (timeout >= 0) {
		ts.tv_sec = timeout / 1000;
		ts.tv_nsec = (timeout % 1000) * 1000000;
		tsp = &ts;
	} else
		tsp = NULL;

	return epoll_pwait2(epfd, events, maxevents, tsp, sigmask);
}
